File: week_1/Data Structure/heap.py
class HeapType:
    def __init__(self):
        self.heap = []
        self.heap_size = 0

def createHeap():
    h = HeapType()
    h.heap_size = 0
    return h

def insertHeap(h, item):
    h.heap_size += 1
    i = h.heap_size
    while i != 1 and item > h.heap[i // 2]:
        h.heap[i] = h.heap[i // 2]
        i //= 2
    h.heap[i] = item

def deleteHeap(h):
    item = h.heap[1]
    temp = h.heap[h.heap_size]
    h.heap_size -= 1
    j = 2
    i = 1
    while j <= h.heap_size and h.heap[j] > temp:
        if j < h.heap_size and h.heap[j] < h.heap[j + 1]:
            j += 1
        if temp >= h.heap[j]:
            break
        else:
            h.heap[i] = h.heap[j]
            i = j
            j = j * 2
    h.heap[i] = temp
    return item

def printHeap(h):
    print("Heap: ", end="")
    for i in range(1, h.heap_size + 1):
        print("[%d] " % h.heap[i], end="")

heap = createHeap()
insertHeap(heap, 10)
insertHeap(heap, 45)
insertHeap(heap, 19)
insertHeap(heap, 11)
insertHeap(heap, 96)
printHeap(heap)
n = heap.heap_size
for i in range(1, n + 1):
    item = deleteHeap(heap)
    print("\n delete: [%d] " % item, end="")

File: week_1/Data Structure/heap.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_ELEMENT 100

typedef struct {
	int heap[MAX_ELEMENT];
	int heap_size;
} heapType;

heapType* createHeap() {
	heapType* h = (heapType*)malloc(sizeof(heapType));
	h->heap_size = 0;
	return h;
}

// 힙에 item을 삽입하는 연산
void insertHeap(heapType* h, int item) {
	int i;
	h->heap_size = h->heap_size + 1;
	i = h->heap_size;

	while ((i != 1) && (item > h->heap[i / 2])) {
		h->heap[i] = h->heap[i / 2];
		i /= 2;
	}
	h->heap[i] = item;
}

// 힙의 루트를 삭제하여 반환하는 연산
int deleteHeap(heapType* h) {
	int item = h->heap[1];
	int temp = h->heap[h->heap_size];
	h->heap_size--;

	int j = 2, i = 1;
	while (j <= h->heap_size && h->heap[j] > temp) {
		if (j < h->heap_size && h->heap[j] < h->heap[j + 1]) j++;
		if (temp >= h->heap[j]) break;
		else {
			h->heap[i] = h->heap[j];
			i = j;
			j = j * 2;
		}
	}
	h->heap[i] = temp;

	return item;
}

void printHeap(heapType* h) {
	int i;
	printf("Heap : ");
	for (i = 1; i <= h->heap_size; i++) {
		printf("[%d] ", h->heap[i]);
	}
}

int main() {

	int i, n, item;
	heapType* heap = createHeap();
	insertHeap(heap, 10);
	insertHeap(heap, 45);
	insertHeap(heap, 19);
	insertHeap(heap, 11);
	insertHeap(heap, 96);

	printHeap(heap);

	n = heap->heap_size;
	for (i = 1; i <= n; i++) {
		item = deleteHeap(heap);
		printf("\n delete : [%d] ", item);
	}
	getchar();
	return 0;
}